File: docs/JAMIA/comparisons/RPython/TM/36462257_empty_df_loses_dtype.ts
/**
 * RPython SO#36462257 — Create Empty Dataframe in Pandas specifying column types
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In pandas, pd.DataFrame() cannot accept per-column dtypes at construction.
 * Users must create empty and then cast, risking schema drift.
 *
 * In tidy-ts, the schema is defined by the row type. Column types are inferred
 * from the data provided. There is no separate "dtype" specification that can
 * diverge from the actual data.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

// Schema is defined by the row shape — types are explicit and consistent
const df = createDataFrame([
  { contract: "H1234", state: "CA", membership: 500, raf: 1.05 },
  { contract: "H5678", state: "TX", membership: 300, raf: 0.92 },
]);

// Numeric columns are number, string columns are string — no ambiguity
console.log(s.mean(df.extract("raf")));

// Attempting numeric ops on string columns is caught at compile time
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("contract"));

File: docs/JAMIA/comparisons/RPython/TM/36462257_empty_df_loses_dtype.py
# ID: SO#36462257
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: Empty DataFrame loses dtype specification. Schema lost through operation.
# Reproduction status: Live
# Type system catch: Schema preserves column types through empty operations

import pandas as pd

# BUG: dtype parameter only accepts a single type, not a list
try:
    df = pd.DataFrame(
        index=['pbp'],
        columns=['contract', 'state', 'membership', 'raf'],
        dtype=['str', 'str', 'int', 'float']
    )
    print(df)
except TypeError as e:
    print(f"Crash: {e}")
